maven-core/src/main/java/org/apache/maven/lifecycle/internal/LifecycleDependencyResolver.java

```diff
@@ -90,18 +89,17 @@ public LifecycleDependencyResolver(ProjectDependenciesResolver projectDependenci
 
     public static List<MavenProject> getProjects(MavenProject project, MavenSession session, boolean aggregator) {
         if (aggregator && project.getCollectedProjects() != null) {
-            return getProjectAndSubModules(project).collect(Collectors.toList());
+            List<MavenProject> downstreamProjects = session.getProjectDependencyGraph()
+                    .getDownstreamProjects(project, true); // sorted but more than needed
+            List<MavenProject> collectedProjects = project.getCollectedProjects(); // not sorted but what we need
```

Comment:
This should be `getProjectAndSubModules(project).collect(Collectors.toList())` instead of `project.getCollectedProjects()`, else the content will be very different.
The first one will return the project and all its submodules (recursively), while the second will just return the modules of the current project.

Reply:
Ah, so only 1 level... right

Reply:
fixed

Reply:
I cannot confirm this statement with 3.8.7: https://github.com/apache/maven-doxia-sitetools/pull/94/files#r1136260250

Reply:
This would mean we don't need to recurse.

Reply:
The question is what is expected here? I debugged with Maven SCM and all modules were in that list.

Reply:
I think I misread the code at https://github.com/apache/maven/blob/8150c62749a637a98fbb9489fb8162034956220b/maven-core/src/main/java/org/apache/maven/project/DefaultProjectBuilder.java#L626-L631

Reading more closely, the `projects` list will contain all submodules recursively...